include/tvm/runtime/micro/utvm_device_lib.h

```diff
@@ -0,0 +1,52 @@
+/*!
+ *  Copyright (c) 2019 by Contributors
+ * \file utvm_device_lib.h
+ * \brief utvm device library definitions
+ */
+#ifndef TVM_RUNTIME_MICRO_UTVM_DEVICE_LIB_H_
+#define TVM_RUNTIME_MICRO_UTVM_DEVICE_LIB_H_
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+#include <stdint.h>
+
+void *(*TVMBackendAllocWorkspace_)(int, int, uint64_t, int, int) =
+    (void *(*)(int, int, uint64_t, int, int)) 1;
+int (*TVMBackendFreeWorkspace_)(int, int, void*) = (int (*)(int, int, void*)) 1;
+void (*TVMAPISetLastError_)(const char*) = (void (*)(const char*)) 1;
+
+void* TVMBackendAllocWorkspace(int device_type, int device_id, uint64_t size,
+    int dtype_code_hint, int dtype_bits_hint) {
+  return (*TVMBackendAllocWorkspace_)(device_type, device_id, size, dtype_code_hint,
+                                      dtype_bits_hint);
+}
+
+int TVMBackendFreeWorkspace(int device_type, int device_id, void* ptr) {
+  return (*TVMBackendFreeWorkspace_)(device_type, device_id, ptr);
+}
+
+void TVMAPISetLastError(const char* msg) {
+  (*TVMAPISetLastError_)(msg);
+}
+
+float min(float a, float b) {
```

Comment:
We can't rename these functions without modifying the C codegen, because `min` and `max` are the names that are generated by the source that includes this header.

As for switching to macros, I would think modern compilers are smart enough to inline small functions such as this.  I could be wrong though.

Reply:
For modern compilers, such functions are defined within certain namespace like `std`, and they typically use `template` based implementation. 

However, this header file is more likely to be a C header. Therefore, following the C convention is more likely to be reasonable. Otherwise, someone who wanted to use `double` might lose precision unintentionally.

Reply:
I think it would be fine to add `#include <math.h>` in `c_runtime_api.h`, and rename `min` to `fmin` in `codegen_c.cc`, and avoid implementing a custom compare function.

Reply:
Adding an include for those functions won't help here, because we don't currently link against libc, so we don't have implementations for those functions.  We have considered loading something like [openlibm](https://github.com/JuliaMath/openlibm) on the device, to get access to math functions, but that would likely be in a future PR.

Reply:
Okay, let's put this temp implement of min/max functions right here.

However, I'm feeling that integration with something like `openlibm` wouldn't be the ultimate goal of TVM. Instead, these functions should be built in TVM, either in the `runtime`, in `topi` or `codegen`. (Actually, they are already been implemented in the later two options.)

Reply:
Yeah.  If there already exist implementations of some of these functions, it might make more sense to modify codegen to inline those implementations in the generated source file.  We'll see what makes sense as we spend more time using MicroTVM.

Reply:
Let us simply modify the CodegenC to directly emit conditional expression ```((x)>(y)? (x):(y))```

Reply:
Will any of the generated expressions contain mutation?  If so, we'll need to create bindings for `x` and `y` to prevent duplicate computations.  If there's no mutation, we can safely inline, but it's still inefficient.

Reply:
@tqchen As far as I understand, there _are_ some operators that are stateful, so while the current codegen modification works for ResNet, I don't know if we want to merge this change in.

Reply:
@liangfu - On the topic about what comes from libm vs what comes from the micro tvm device run time we need to think about where to draw the line with what comes in the C runtimes from static toolchains that we would be using to compile this. For instance if this is to be using a variant of a baremetal gnu toolchain as has been suggested (to compile the generated C backend file), the library of choice these days for baremetal stuff is actually newlib (and newlib-nano) and thus you get some of the libm functions for free in this regard. Putting in all traditional math runtime functions in  a micro tvm runtime sounds like the wrong call to me as you'd need to get this anyway from the static toolchain that was being used to link things in.

On the topic of macros vs inlines, if you put these out as static always_inline functions you get the same effect, inlining works also at -O0. In fact this is a common technique used in various intrinsic header files that are provided by GCC to force the inlining for intrinsics, thus you could avoid macros.

I hope this helps.

Reply:
@u99127 It comes more clear to me that C backend is very much different from LLVM backend, which generates instruction-level optimized code directly. And I agree that integrating with either `newlib` or `newlib-nano` should be a preferable choice to complement the baremetal C runtime. However, IMHO, the C codegen should be designed for C (with some compatibility with C++), not exclusively for C++. For instance, we should avoid using `max` for both `float` and `double`, instead, `fmaxf` and `fmax` should be used (, even in this temp implement of the C math library).

On the macros vs inlines topic, I agree that inlines is preferred over macros. For instance, `MIN(MAX(a, b), c)` can be simply expanded into `(a>b?a:b)<c?(a>b?a:b):c`, therefore, the comparison between `a` and `b` could have been performed twice.

Reply:
I still think that we should directly implement it as a conditional expression in the C codegen. At the moment there is no problem of stateful expressions. If we do in the future, they will be separated into a different statement, which would not affect the semantics.

To be more careful in the future, we can just print a new variable assignment for each sub expression. and then refer to the variable. see https://github.com/dmlc/tvm/blob/cffb4fba03ea582417e2630bd163bca773756af6/src/codegen/codegen_c.cc#L129